netbsd/Platform.c

```diff
@@ -0,0 +1,324 @@
+   if (err) {
+      *one = 0;
+      *five = 0;
+      *fifteen = 0;
+      return;
+   }
+   *one     = (double) loadAverage.ldavg[0] / loadAverage.fscale;
+   *five    = (double) loadAverage.ldavg[1] / loadAverage.fscale;
+   *fifteen = (double) loadAverage.ldavg[2] / loadAverage.fscale;
+}
+
+int Platform_getMaxPid() {
+   // this is hard-coded in sys/sys/proc.h - no sysctl exists
+   return 30000;
+}
+
+double Platform_setCPUValues(Meter* this, int cpu) {
+   const NetBSDProcessList* pl = (const NetBSDProcessList*) this->pl;
+   const CPUData* cpuData = &(pl->cpus[cpu]);
+   double total = cpuData->totalPeriod == 0 ? 1 : cpuData->totalPeriod;
+   double totalPercent;
+   double* v = this->values;
+
+   v[CPU_METER_NICE] = cpuData->nicePeriod / total * 100.0;
+   v[CPU_METER_NORMAL] = cpuData->userPeriod / total * 100.0;
+   if (this->pl->settings->detailedCPUTime) {
+      v[CPU_METER_KERNEL]  = cpuData->sysPeriod / total * 100.0;
+      v[CPU_METER_IRQ]     = cpuData->intrPeriod / total * 100.0;
+      v[CPU_METER_SOFTIRQ] = 0.0;
+      v[CPU_METER_STEAL]   = 0.0;
+      v[CPU_METER_GUEST]   = 0.0;
+      v[CPU_METER_IOWAIT]  = 0.0;
+      v[CPU_METER_FREQUENCY] = NAN;
+      this->curItems = 8;
+      totalPercent = v[0] + v[1] + v[2] + v[3];
+   } else {
+      v[2] = cpuData->sysAllPeriod / total * 100.0;
+      v[3] = 0.0; // No steal nor guest on NetBSD
+      totalPercent = v[0] + v[1] + v[2];
+      this->curItems = 4;
+   }
+
+   totalPercent = CLAMP(totalPercent, 0.0, 100.0);
+
+   v[CPU_METER_TEMPERATURE] = NAN;
+
+   return totalPercent;
+}
+
+void Platform_setMemoryValues(Meter* this) {
+   const ProcessList* pl = this->pl;
+   long int usedMem = pl->usedMem;
+   long int buffersMem = pl->buffersMem;
+   long int cachedMem = pl->cachedMem;
+   usedMem -= buffersMem + cachedMem;
+   this->total = pl->totalMem;
+   this->values[0] = usedMem;
+   this->values[1] = buffersMem;
+   this->values[2] = cachedMem;
+}
+
+void Platform_setSwapValues(Meter* this) {
+   const ProcessList* pl = this->pl;
+   this->total = pl->totalSwap;
+   this->values[0] = pl->usedSwap;
+   this->values[1] = NAN;
+}
+
+char* Platform_getProcessEnv(pid_t pid) {
+   char errbuf[_POSIX2_LINE_MAX];
+   char* env;
+   char** ptr;
+   int count;
+   kvm_t* kt;
+   struct kinfo_proc* kproc;
+   size_t capacity = 4096, size = 0;
+
+   if ((kt = kvm_openfiles(NULL, NULL, NULL, KVM_NO_FILES, errbuf)) == NULL) {
+      return NULL;
+   }
+
+   if ((kproc = kvm_getprocs(kt, KERN_PROC_PID, pid, &count)) == NULL) {
+      (void) kvm_close(kt);
+      return NULL;
+   }
+
+   if ((ptr = kvm_getenvv(kt, kproc, 0)) == NULL) {
+      (void) kvm_close(kt);
+      return NULL;
+   }
+
+   env = xMalloc(capacity);
+   for (char** p = ptr; *p; p++) {
+      size_t len = strlen(*p) + 1;
+
+      if (size + len > capacity) {
+         capacity *= 2;
+         env = xRealloc(env, capacity);
+      }
+
+      strlcpy(env + size, *p, len);
```

Comment:
What about using `String_safeStrncpy` instead?

Reply:
Should be fixed in ed2f0dd